signer/kms/gcp/client.go

```diff
@@ -0,0 +1,399 @@
+	AlgorithmRSAPSS2048SHA256      = "rsa-pss-2048-sha256"
+	AlgorithmRSAPSS3072SHA256      = "rsa-pss-3072-sha256"
+	AlgorithmRSAPSS4096SHA256      = "rsa-pss-4096-sha256"
+	AlgorithmRSAPSS4096SHA512      = "rsa-pss-4096-sha512"
+)
+
+var algorithmMap = map[string]kmspb.CryptoKeyVersion_CryptoKeyVersionAlgorithm{
+	AlgorithmECDSAP256SHA256:       kmspb.CryptoKeyVersion_EC_SIGN_P256_SHA256,
+	AlgorithmECDSAP384SHA384:       kmspb.CryptoKeyVersion_EC_SIGN_P384_SHA384,
+	AlgorithmRSAPKCS1v152048SHA256: kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_2048_SHA256,
+	AlgorithmRSAPKCS1v153072SHA256: kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_3072_SHA256,
+	AlgorithmRSAPKCS1v154096SHA256: kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_4096_SHA256,
+	AlgorithmRSAPKCS1v154096SHA512: kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_4096_SHA512,
+	AlgorithmRSAPSS2048SHA256:      kmspb.CryptoKeyVersion_RSA_SIGN_PSS_2048_SHA256,
+	AlgorithmRSAPSS3072SHA256:      kmspb.CryptoKeyVersion_RSA_SIGN_PSS_3072_SHA256,
+	AlgorithmRSAPSS4096SHA256:      kmspb.CryptoKeyVersion_RSA_SIGN_PSS_4096_SHA256,
+	AlgorithmRSAPSS4096SHA512:      kmspb.CryptoKeyVersion_RSA_SIGN_PSS_4096_SHA512,
+}
+
+var (
+	errKMSReference = errors.New("kms specification should be in the format gcpkms://projects/[PROJECT_ID]/locations/[LOCATION]/keyRings/[KEY_RING]/cryptoKeys/[KEY]/cryptoKeyVersions/[VERSION]")
+
+	re = regexp.MustCompile(`^gcpkms://projects/([^/]+)/locations/([^/]+)/keyRings/([^/]+)/cryptoKeys/([^/]+)(?:/(?:cryptoKeyVersions|versions)/([^/]+))?$`)
+)
+
+// ReferenceScheme schemes for various KMS services are copied from https://github.com/google/go-cloud/tree/master/secrets
+const ReferenceScheme = "gcpkms://"
+
+// ValidReference returns a non-nil error if the reference string is invalid
+func ValidReference(ref string) error {
+	if !re.MatchString(ref) {
+		return errKMSReference
+	}
+	return nil
+}
+
+func parseReference(resourceID string) (projectID, locationID, keyRing, keyName, version string, err error) {
+	v := re.FindStringSubmatch(resourceID)
+	if len(v) != 6 {
+		err = fmt.Errorf("invalid gcpkms format %q", resourceID)
+		return
+	}
+	projectID, locationID, keyRing, keyName, version = v[1], v[2], v[3], v[4], v[5]
+	return
+}
+
+type gcpClient struct {
+	projectID  string
+	locationID string
+	keyRing    string
+	keyName    string
+	version    string
+	kvCache    *ttlcache.Cache[string, cryptoKeyVersion]
+	client     *gcpkms.KeyManagementClient
+}
+
+func newGCPClient(ctx context.Context, ksp *kms.KMSSignerProvider) (*gcpClient, error) {
+	if err := ValidReference(ksp.Reference); err != nil {
+		return nil, err
+	}
+
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	g := &gcpClient{
+		kvCache: nil,
+	}
+
+	var err error
+	g.projectID, g.locationID, g.keyRing, g.keyName, g.version, err = parseReference(ksp.Reference)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: We need to think about how we want to handle passing in options here
+	g.client, err = gcpkms.NewKeyManagementClient(ctx)
```

Comment:
We can use the functional option pattern we use elsewhere to supply KeyManagementClientOptions

```
func WithKeyManagementClientOptions(opts ...gcpkms.ClientOption) Option {
    return func() { ... }
}
```

We can pair up on this if you like :)

Reply:
yeah so my NOTE here is more along the lines of making sure how we consider the UX of presenting the options in the CLI. I am just wanting to avoid the `--everything` dilemma. Do you think that environment variables would be an acceptable way of exposing options?